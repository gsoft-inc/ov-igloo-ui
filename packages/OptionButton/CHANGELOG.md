# @igloo-ui/option-button

## 0.1.2

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency

## 0.1.1

### Patch Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

## 0.1.0

### Minor Changes

- 24c2194: Rebrand of OptionButton

## 0.0.5

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 0.0.4

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.0.3

### Patch Changes

- 0640409: udpate classnames to 2.3.2

## 0.0.2

### Patch Changes

- d971143: All tests were updated to use testing library and not enzyme

## 0.0.1

### Patch Changes

- dad5e79: Feat: Initial release of the OptionButton component
