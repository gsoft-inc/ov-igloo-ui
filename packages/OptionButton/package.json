{
  "name": "@igloo-ui/option-button",
  "version": "0.0.5",
  "main": "dist/OptionButton.js",
  "module": "dist/OptionButton.js",
  "types": "dist/OptionButton.d.ts",
  "styles": "dist/option-button.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "classnames": "^2.3.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
