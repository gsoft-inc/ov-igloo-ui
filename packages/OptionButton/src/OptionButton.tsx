import * as React from 'react';
import cx from 'classnames';

import {
  TextIcon,
  OptionScaleIcon,
  MultipleChoiceIcon,
  LikertIcon,
  TextIconDisabled,
  OptionScaleIconDisabled,
  MultipleChoiceIconDisabled,
  LikertIconDisabled,
} from './svgs';

import './option-button.scss';

export type ButtonType = 'Text' | 'OptionScale' | 'MultipleChoice' | 'Likert';

export interface OptionButtonProps extends React.ComponentProps<'input'> {
  /** True when the option button is selected */
  checked?: boolean;
  /** Add a specific class to the option button */
  className?: string;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Add a description that is displayed below the text prop */
  description?: string;
  /** True if the option button should be disabled */
  disabled?: boolean;
  /** Add an icon the the beginning of the option button.
   * This overrides the icon that comes with buttonType */
  icon?: React.ReactNode;
  /** Add an id to the radio button of the option button.
   * This will also be used as the htmlFor attribute on the label */
  id: string;
  /** Add an onChange event to the option button */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  /** Added the main text to the option button */
  text: string;
  /** Add a button type, which will come with its own icon.
   * If adding an icon, this will be overridden */
  buttonType?: ButtonType;
  /** True if the option button should have an unchecked appearance */
  unchecked?: boolean;
}

const OptionButton: React.FunctionComponent<OptionButtonProps> = (
  props: OptionButtonProps
) => {
  const {
    checked,
    className,
    dataTest,
    description,
    disabled,
    icon,
    id,
    onChange,
    text,
    buttonType,
    unchecked,
    ...rest
  } = props;

  const renderIcon = (): React.ReactNode => {
    if (icon) {
      return icon;
    }
    switch (buttonType) {
      case 'Text':
        if (disabled || unchecked) {
          return (
            // eslint-disable-next-line max-len
            <TextIconDisabled className="ids-option-button__text-icon-disabled" />
          );
        }
        return <TextIcon className="ids-option-button__text-icon" />;
      case 'OptionScale':
        if (disabled || unchecked) {
          return (
            // eslint-disable-next-line max-len
            <OptionScaleIconDisabled className="ids-option-button__option-scale-icon-disabled" />
          );
        }
        return (
          // eslint-disable-next-line max-len
          <OptionScaleIcon className="ids-option-button__option-scale-icon" />
        );
      case 'MultipleChoice':
        if (disabled || unchecked) {
          return (
            // eslint-disable-next-line max-len
            <MultipleChoiceIconDisabled className="ids-option-button__multiple-choice-icon-disabled" />
          );
        }
        return (
          // eslint-disable-next-line max-len
          <MultipleChoiceIcon className="ids-option-button__multiple-choice-icon" />
        );
      case 'Likert':
        if (disabled || unchecked) {
          return (
            // eslint-disable-next-line max-len
            <LikertIconDisabled className="ids-option-button__likert-icon-disabled" />
          );
        }
        return <LikertIcon className="ids-option-button__likert-icon" />;
      default:
        return null;
    }
  };

  const classes = cx('ids-option-button', className, {
    'ids-option-button--unchecked': unchecked,
  });

  return (
    <span className={classes} data-test={dataTest}>
      <input
        id={id}
        className="ids-option-button__radio"
        onChange={onChange}
        disabled={disabled}
        checked={checked}
        type="radio"
        {...rest}
      />
      <label htmlFor={id} className="ids-option-button__label">
        <span className="ids-option-button__icon-container">
          {renderIcon()}
        </span>
        <span className="ids-option-button__text-container">
          <span className="ids-option-button__text">{text}</span>
          <span className="ids-option-button__desc">{description}</span>
        </span>
      </label>
    </span>
  );
};

export default OptionButton;
