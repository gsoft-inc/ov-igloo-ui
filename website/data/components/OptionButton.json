[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/OptionButton/src/OptionButton.tsx",
      "description": "",
      "displayName": "OptionButton",
      "methods": [],
      "props": {
        "buttonType": {
          "defaultValue": null,
          "description": "Add a button type, which will come with its own icon.\nIf adding an icon, this will be overridden",
          "name": "buttonType",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "ButtonType" }
        },
        "checked": {
          "defaultValue": null,
          "description": "True when the option button is selected",
          "name": "checked",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "children": {
          "defaultValue": null,
          "description": "Add the main text to the option button",
          "name": "children",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            },
            {
              "fileName": "components/node_modules/@types/react/index.d.ts",
              "name": "TypeLiteral"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "className": {
          "defaultValue": null,
          "description": "Add a specific class to the option button",
          "name": "className",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "description": {
          "defaultValue": null,
          "description": "Add a description that is displayed below the text prop",
          "name": "description",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "disabled": {
          "defaultValue": null,
          "description": "True if the option button should be disabled",
          "name": "disabled",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "htmlFor": {
          "defaultValue": null,
          "description": "Add an htmlFor to the radio button of the option button.\nThis will also be used as the htmlFor attribute on the label",
          "name": "htmlFor",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": true,
          "type": { "name": "string" }
        },
        "icon": {
          "defaultValue": null,
          "description": "Add an icon to the beginning of the option button.\nThis overrides the icon that comes with buttonType",
          "name": "icon",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "onChange": {
          "defaultValue": null,
          "description": "Add an onChange event to the option button",
          "name": "onChange",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "(e: ChangeEvent<HTMLInputElement>) => void" }
        },
        "unchecked": {
          "defaultValue": null,
          "description": "True if the option button should have an unchecked appearance",
          "name": "unchecked",
          "parent": {
            "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
            "name": "OptionButtonProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/OptionButton/src/OptionButton.tsx",
              "name": "OptionButtonProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        }
      }
    }
  ]
]
