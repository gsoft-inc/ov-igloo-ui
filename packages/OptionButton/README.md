# OptionButton

The option button is a component that will often be used in a group of multiple option buttons and will play a role similar to the radio buttons.

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/option-button` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/option-button
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/option-button
```

## Usage

Then to use the component in your code just import it!

```jsx
import OptionButton from '@igloo-ui/option-button';

<OptionButton id="textBtn" name="buttonType" buttonType="Text" text="Text" />;
```
