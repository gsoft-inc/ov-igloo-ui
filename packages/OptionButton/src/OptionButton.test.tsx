/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';

import OptionButton, { OptionButtonProps, ButtonType } from './OptionButton';
import LabelSolid from '@igloo-ui/icons/dist/LabelSolid';

const setup = (props: OptionButtonProps) => {
  return shallow(<OptionButton {...props} />);
};

describe('OptionButton', () => {
  const defaultComponentSetUp: OptionButtonProps = {
    children: 'I am an option button',
    htmlFor: 'option-button-test-1',
    dataTest: 'option-button-test-1',
  };
  const component = () => setup(defaultComponentSetUp);
  test('It should render without errors', () => {
    const wrapper = component().find('.ids-option-button');
    expect(wrapper.length).toBe(1);
  });

  const expectToHaveProperIcon = (
    buttonType: ButtonType,
    disabled: boolean
  ) => {
    const props = {
      children: 'option button',
      htmlFor: 'option-button-test-5',
      buttonType: buttonType,
      disabled: disabled,
    };
    let optionButtonClassName = '';
    render(<OptionButton {...props} />);
    switch (buttonType) {
      case 'optionScale':
        if (disabled) {
          optionButtonClassName =
            'ids-option-button__option-scale-icon disabled';
        }
        optionButtonClassName = 'ids-option-button__option-scale-icon';
        break;
      case 'multipleChoice':
        if (disabled) {
          optionButtonClassName =
            'ids-option-button__multiple-choice-icon disabled';
        }
        optionButtonClassName = 'ids-option-button__multiple-choice-icon';
        break;
      case 'likert':
        if (disabled) {
          optionButtonClassName = 'ids-option-button__likert-icon disabled';
        }
        optionButtonClassName = 'ids-option-button__likert-icon';
        break;
      default:
        if (disabled) {
          optionButtonClassName = 'ids-option-button__text-icon disabled';
        }
        optionButtonClassName = 'ids-option-button__text-icon';
        break;
    }

    const svg = document.getElementsByClassName(optionButtonClassName);
    expect(svg).toBeTruthy();
  };

  test('It should render a snapshot', () => {
    expect(component()).toMatchSnapshot();
  });

  test('It should render a checked state', () => {
    const { getByRole } = render(
      <OptionButton htmlFor="option-button-test-2" checked onChange={() => {}}>
        Hello World!
      </OptionButton>
    );
    const radio = getByRole('radio');
    expect(radio).toBeChecked();
  });

  test('It should render a disabled state', () => {
    const { getByRole } = render(
      <OptionButton htmlFor="option-button-test-3" disabled onChange={() => {}}>
        Hello World!
      </OptionButton>
    );
    const radio = getByRole('radio');
    expect(radio).toBeDisabled();
  });

  test('It should render a unchecked look', () => {
    const wrapper = setup({ ...defaultComponentSetUp, unchecked: true });
    const radio = wrapper.find('.ids-option-button--unchecked');
    expect(radio.length).toBe(1);
  });

  test('It should render an option button with an icon', () => {
    render(
      <OptionButton
        htmlFor="option-button-test-4"
        icon={<LabelSolid size="small" />}
      >
        I am an option button with an icon
      </OptionButton>
    );
    const svg = document.getElementsByTagName('svg');
    expect(svg.length).toBe(1);
  });

  test('It should render an option button with the proper icon', () => {
    expectToHaveProperIcon('likert', false);
    expectToHaveProperIcon('likert', true);
    expectToHaveProperIcon('multipleChoice', false);
    expectToHaveProperIcon('multipleChoice', true);
    expectToHaveProperIcon('optionScale', false);
    expectToHaveProperIcon('optionScale', true);
    expectToHaveProperIcon('text', false);
    expectToHaveProperIcon('text', true);
  });
});
