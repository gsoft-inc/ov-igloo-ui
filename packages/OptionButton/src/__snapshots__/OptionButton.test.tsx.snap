// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionButton It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-option-button"
    data-test="option-button-test-1"
  >
    <input
      class="ids-option-button__radio"
      id="option-button-test-1"
      type="radio"
    />
    <label
      class="ids-option-button__label"
      for="option-button-test-1"
    >
      <span
        class="ids-option-button__icon-container"
      >
        <svg
          class="ids-option-button__text-icon "
          fill="none"
          height="32"
          viewBox="0 0 32 32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#clip0_5250_851)"
          >
            <circle
              cx="16"
              cy="16"
              fill="#B2DDF2"
              r="16"
            />
            <path
              d="M23.963 26.0296C25.03 26.0296 24.2124 10.0177 24.0067 6.27549C23.9825 5.83494 23.6247 5.5 23.1834 5.5L8.65899 5.50005C6.65998 5.50007 7.85352 22.788 7.85352 25.3092C7.85352 27.4703 22.7548 26.0296 23.963 26.0296Z"
              fill="white"
            />
            <path
              clip-rule="evenodd"
              d="M15.5358 11.2583C13.4703 11.2877 11.5563 11.3885 11.1234 11.4607C10.7148 11.5288 10.3284 11.2527 10.2603 10.8442C10.1922 10.4356 10.4682 10.0492 10.8768 9.98107C11.4439 9.88655 13.48 9.78738 15.5144 9.75844C16.5465 9.74375 17.6038 9.74678 18.5082 9.77915C19.3876 9.81064 20.1947 9.87143 20.682 9.99325C21.0839 10.0937 21.3282 10.5009 21.2277 10.9028C21.1273 11.3046 20.7201 11.5489 20.3182 11.4485C20.0055 11.3703 19.3501 11.3103 18.4545 11.2782C17.584 11.247 16.5537 11.2438 15.5358 11.2583Z"
              fill="#233043"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M20.3768 16.2398C17.4672 15.7549 15.043 15.7505 11.0931 16.2442C10.6821 16.2956 10.3072 16.004 10.2559 15.593C10.2045 15.182 10.496 14.8071 10.9071 14.7558C14.9571 14.2495 17.533 14.2452 20.6234 14.7602C21.0319 14.8283 21.308 15.2147 21.2399 15.6233C21.1718 16.0319 20.7854 16.3079 20.3768 16.2398Z"
              fill="#233043"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M14.7057 20.2437C13.0994 20.2798 11.5633 20.3874 11.147 20.4716C10.741 20.5537 10.3453 20.2912 10.2632 19.8852C10.1811 19.4792 10.4436 19.0835 10.8496 19.0014C11.4189 18.8862 13.0906 18.7796 14.672 18.7441C15.4776 18.7259 16.2896 18.7256 16.9636 18.754C17.3 18.7682 17.6116 18.7899 17.8755 18.822C18.1196 18.8516 18.39 18.8972 18.6031 18.9843C18.9865 19.1412 19.1701 19.5791 19.0133 19.9625C18.8565 20.3458 18.4186 20.5295 18.0352 20.3727C18.0356 20.3729 18.0355 20.3728 18.0348 20.3726C18.0305 20.3711 18.0042 20.3623 17.9462 20.3503C17.883 20.3372 17.7996 20.3238 17.6947 20.311C17.4846 20.2855 17.2158 20.266 16.9004 20.2527C16.2708 20.2261 15.494 20.226 14.7057 20.2437Z"
              fill="#233043"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M8.27614 12.5686C8.29077 13.3727 8.30604 14.2119 8.30606 15.0696C8.30609 17.4809 8.30618 19.7263 8.36854 21.5679C8.42383 23.2007 8.52714 24.4356 8.69522 25.1787C8.81252 25.2193 8.98242 25.2678 9.21711 25.318C9.64611 25.4096 10.2034 25.4893 10.8607 25.5543C12.1731 25.684 13.8294 25.75 15.5429 25.75C17.2552 25.75 19.0097 25.6841 20.517 25.5534C21.8058 25.4417 22.864 25.2863 23.5583 25.1018C23.5865 24.9612 23.6163 24.7685 23.6429 24.515C23.6938 24.029 23.7258 23.402 23.7405 22.6606C23.77 21.1795 23.7297 19.2894 23.6482 17.2528C23.4972 13.4812 23.2061 9.24315 22.9646 6.25H9.06538C8.97554 6.25 8.92757 6.27219 8.88543 6.30522C8.83102 6.34787 8.75019 6.44003 8.66434 6.63052C8.48576 7.02682 8.3659 7.66299 8.30421 8.54296C8.22378 9.69028 8.24885 11.0682 8.27614 12.5686ZM7.96007 5.12467C8.27561 4.87733 8.65213 4.75 9.06538 4.75H23.6555C24.0455 4.75 24.3705 5.04893 24.4029 5.43759C24.6551 8.4571 24.9829 13.0938 25.147 17.1928C25.229 19.241 25.2705 21.1654 25.2402 22.6904C25.2251 23.4519 25.1918 24.1258 25.1347 24.6713C25.0809 25.1847 24.9977 25.6827 24.8366 26.0233C24.751 26.2042 24.5966 26.3434 24.4078 26.4099C23.5554 26.7102 22.1821 26.9147 20.6466 27.0478C19.0881 27.183 17.2895 27.25 15.5429 27.25C13.7975 27.25 12.0893 27.1831 10.7132 27.047C10.0263 26.9791 9.40832 26.8926 8.90382 26.7849C8.43294 26.6843 7.95202 26.5445 7.62722 26.3157C7.49288 26.221 7.39334 26.0849 7.34391 25.9282C7.05634 25.0165 6.93167 23.4577 6.8694 21.6187C6.80613 19.75 6.80609 17.4817 6.80606 15.084L6.80606 15.0696C6.80604 14.285 6.79168 13.4831 6.77757 12.6947C6.74983 11.1454 6.72302 9.64853 6.80788 8.43806C6.87199 7.52372 7.0037 6.66464 7.29679 6.01425C7.44678 5.68141 7.65681 5.36238 7.96007 5.12467Z"
              fill="#233043"
              fill-rule="evenodd"
            />
          </g>
          <defs>
            <clippath
              id="clip0_5250_851"
            >
              <rect
                fill="white"
                height="32"
                width="32"
              />
            </clippath>
          </defs>
        </svg>
      </span>
      <span
        class="ids-option-button__text-container"
      >
        <span
          class="ids-option-button__text"
        >
          I am an option button
        </span>
        <span
          class="ids-option-button__desc"
        />
      </span>
    </label>
  </div>
</DocumentFragment>
`;
